include/envoy/stats/symbol_table.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <vector>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Stats {
+/**
+ * Interface for shortening and retrieving stat names.
+ * Guarantees that x = decode(encode(x)) for any x.
+ *
+ * Even though the symbol table does manual reference counting, curr_counter_ is monotonically
+ * increasing. So encoding "foo", freeing the sole stat containing "foo", and then re-encoding "foo"
```

Comment:
If you have 100 clusters updated every 1s, this is 497 days. We probably have less frequent cluster updates but some deployments will be larger. I don't mind the idea of a free list to avoid this. I'm not fully sold on efficiency here of vector vs. map, it's a time-space tradeoff.